### PR TITLE
Glutin 0.21

### LINKED ISSUE #23189: Update to Glutin 0.21.0
https://github.com/servo/servo/blob/7b8a898ac2ffef5ced8202dfd8bb40e37b972334/ports/servo/Cargo.toml#L49

Hey folks. Glutin v0.21.0 is out.

### PR DESCRIPTION
**Pull request by paulrouget**
Fix #23189

Depends on #23233

Dependencies update:
- https://github.com/servo/rust-webvr/pull/73
- https://github.com/servo/skia/pull/172
- https://github.com/servo/rust-glx/pull/23
- https://github.com/servo/rust-offscreen-rendering-context/pull/134

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23288)

### CODE REVIEW

File: components/canvas/Cargo.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ compositing = {path = "../compositing"}
 cssparser = "0.25"
 euclid = "0.19"
 fnv = "1.0"
-gleam = "0.6.4"
+gleam = "^0.6.7"
```
Comment (jdm) [inline components/canvas/Cargo.toml]: The ^ shouldn't be necessary in any of these changes.

**Review by jdm — APPROVED** [review]

**Comment by Manishearth** [comment]
Also depends on https://github.com/servo/rust-azure/pull/298

**Comment by Manishearth** [comment]
Needs a bump of azure to v0.36.0

**Comment by paulrouget** [comment]
Running into a compilation issue with rand_core…

**Comment by Manishearth** [comment]
Probably because of the multiple rand packages conflicting

**Comment by paulrouget** [comment]
That was not an issue with the RC:
- https://github.com/servo/servo/compare/master...paulrouget:glutin-update-rc

Investigating.

**Comment by paulrouget** [comment]
```
error[E0277]: the trait bound `rand_hc::Hc128Rng: rand_core::SeedableRng` is not satisfied
  --> /Users/paul/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.6.5/src/rngs/std.rs:55:6
   |
55 | impl SeedableRng for StdRng {
   |      ^^^^^^^^^^^ the trait `rand_core::SeedableRng` is not implemented for `rand_hc::Hc128Rng`

error[E0277]: the trait bound `rand_chacha::ChaChaRng: rand_core::SeedableRng` is not satisfied
   --> /Users/paul/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.6.5/src/deprecated.rs:141:6
    |
141 | impl SeedableRng for ChaChaRng {
    |      ^^^^^^^^^^^ the trait `rand_core::SeedableRng` is not implemented for `rand_chacha::ChaChaRng`

error[E0277]: the trait bound `rand_hc::Hc128Rng: rand_core::SeedableRng` is not satisfied
   --> /Users/paul/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.6.5/src/deprecated.rs:198:6
```
and more like this.

**Comment by jdm** [comment]
When I've seen errors like that,bit usually requires updating an older version of the Rand package.

**Comment by paulrouget** [comment]
@jdm it worked. Thank you.

PR updated.

**Comment by paulrouget** [comment]
The way context are handled in this new version of glutin is a bit tricky.

A context is either current or not, and switching from one to the other consumes the context. So we can't simply store a context. See https://github.com/servo/servo/pull/23288/commits/76718f105455c2e4ca644f381b58ec709b800948#diff-7f61eab416c6c9fd2e6bfc9bb5b22f9bR22

**Comment by paulrouget** [comment]
Not ready for review. Need tries and https://github.com/servo/skia/pull/175

**Comment by Manishearth** [comment]
@bors-servo retry

A WPT crash that may be related to glutin, but I'm not sure.

<details>

```
  ▶ CRASH [expected OK] /html/browsers/browsing-the-web/read-media/pageload-image.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ 
  │ (servo:84501): GStreamer-WARNING **: External plugin loader failed. This most likely means that the plugin loader helper binary was not found or could not be run. You might need to set the GST_PLUGIN_SCANNER environment variable if your setup is unusual. This should normally not be required though.
  │ 0:00:00.017259193 84501 0x555d468ee950 WARN      GST_PLUGIN_LOADING gstplugin.c:792:_priv_gst_plugin_load_file_for_registry: module_open failed: libasound.so.2: cannot open shared object file: No such file or directory
  │ 
  │ (servo:84501): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstalsa.so': libasound.so.2: cannot open shared object file: No such file or directory
  │ 0:00:00.017639698 84501 0x555d468ee950 WARN      GST_PLUGIN_LOADING gstplugin.c:792:_priv_gst_plugin_load_file_for_registry: module_open failed: libmms.so.0: cannot open shared object file: No such file or directory
  │ 
  │ (servo:84501): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstmms.so': libmms.so.0: cannot open shared object file: No such file or directory
  │ 0:00:00.017829337 84501 0x555d468ee950 WARN      GST_PLUGIN_LOADING gstplugin.c:792:_priv_gst_plugin_load_file_for_registry: module_open failed: libfaad.so.2: cannot open shared object file: No such file or directory
  │ 
  │ (servo:84501): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstfaad.so': libfaad.so.2: cannot open shared object file: No such file or directory
  │ alloc factor 0.010000 0.900000
  │ alloc factor 0.010000 0.900000
  │ called `Result::unwrap()` on an `Err` value: Io(Custom { kind: ConnectionReset, error: StringError("All senders for this socket closed") }) (thread main, at src/libcore/result.rs:999)
  │ stack backtrace:
  │    0:     0x555d439ca04d - backtrace::backtrace::trace::h1eccb93d7cf3eb11
  │    1:     0x555d439c8eb2 - backtrace::capture::Backtrace::new::hf91afa2b0aa7ec1c
  │    2:     0x555d410dcabf - servo::main::{{closure}}::h50fc25f9586587b0
  │    3:     0x555d4452cc48 - rust_panic_with_hook
  │                         at src/libstd/panicking.rs:478
  │    4:     0x555d4452c6e1 - continue_panic_fmt
  │                         at src/libstd/panicking.rs:381
  │    5:     0x555d4452c5c5 - rust_begin_unwind
  │    6:     0x555d4454b50c - panic_fmt
  │                         at src/libcore/panicking.rs:85
  │    7:     0x555d43a393a2 - core::result::unwrap_failed::hfed186f3252f89a7
  │    8:     0x555d43a1ea77 - webrender_api::api::RenderApi::get_scroll_node_state::h9e34ccd6a0e55700
  │    9:     0x555d4108cb2e - compositing::compositor::IOCompositor<Window>::send_viewport_rects::ha6c144d40408a87f
  │   10:     0x555d4108e692 - _ZN11compositing10compositor26IOCompositor$LT$Window$GT$22handle_browser_message17h80b724a016e4773cE.llvm.8942475133950533092
  │   11:     0x555d4108b50d - compositing::compositor::IOCompositor<Window>::receive_messages::h95323d51e3230b53
  │   12:     0x555d410c5fad - servo::Servo<Window>::handle_events::h001c0ac93015b5e9
  │   13:     0x555d41116b87 - servo::app::App::handle_events::h6f0c52284c78c098
  │   14:     0x555d410f967b - servo::events_loop::EventsLoop::run_forever::hd338dd2909120583
  │   15:     0x555d411159ca - servo::app::App::run::h880381c584067134
  │   16:     0x555d410dc61e - servo::main::ha8485eadd4fc19d7
  │   17:     0x555d410ceed2 - std::rt::lang_start::{{closure}}::h3e338f140d4d4651
  │   18:     0x555d4452c562 - {{closure}}
  │                         at src/libstd/rt.rs:49
  │                          - do_call<closure,i32>
  │                         at src/libstd/panicking.rs:293
  │   19:     0x555d44536e89 - __rust_maybe_catch_panic
  │                         at src/libpanic_unwind/lib.rs:85
  │   20:     0x555d4452d12c - try<i32,closure>
  │                         at src/libstd/panicking.rs:272
  │                          - catch_unwind<closure,i32>
  │                         at src/libstd/panic.rs:388
  │                          - lang_start_internal
  │                         at src/libstd/rt.rs:48
  │   21:     0x555d410dcc81 - main
  │   22:     0x7f02541c782f - __libc_start_main
  │   23:     0x555d4106a378 - _start
  │   24:                0x0 - <unknown>
  └ [2019-05-03T13:04:29Z ERROR servo] called `Result::unwrap()` on an `Err` value: Io(Custom { kind: ConnectionReset, error: StringError("All senders for this socket closed") })
```

</details>

**Comment by jdm** [comment]
That's #13479.

**Comment by Manishearth** [comment]
```
  = note: /Users/servo/buildbot/slave/android-mac/build/android-toolchains/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: fatal error: out of file descriptors and couldn't close any
          collect2: error: ld returned 1 exit status
          toolchain: /Users/servo/buildbot/slave/android-mac/build/android-toolchains/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin
          libs dir: /Users/servo/buildbot/slave/android-mac/build/android-toolchains/ndk/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a
          sysroot: /Users/servo/buildbot/slave/android-mac/build/android-toolchains/ndk/platforms/android-18/arch-arm
          targetdir: /Users/servo/buildbot/slave/android-mac/build/android-toolchains/ndk/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a
```

out of fds 

@bors-servo retry

**Comment by Manishearth** [comment]
same error.


other PRs are not hitting this, maybe this update specifically is causing file descriptor problems? Are we just hitting some limits on `ld`? (I was experiencing other link errors locally with android stuff when working on my XR input PR)

**Comment by jdm** [comment]
I tried restarting the two affected mac machines.

**Comment by jdm** [comment]
New mac builder, same error. This is very weird.

**Comment by paulrouget** [comment]
I'm not sure what I can do here… split the PR in chunks see what triggers the error?

**Comment by jdm** [comment]
That might help, yes.

**Comment by jdm** [comment]
We hit the same android mac build failure in #23392, so it's probably not something special about the glutin upgrade.

**Comment by paulrouget** [comment]
@jdm so what do you think we should do here?

**Comment by jdm** [comment]
Let's again with the full set of changes and see if anything's different. I cleaned out a lot of stale test processes from the mac builders last night.

**Comment by paulrouget** [comment]
Depends on https://github.com/servo/mozangle/pull/19

**Comment by paulrouget** [comment]
I'll see if we can land this with 2 versions of gl_generator (which won't require a mozangle update). And we will de-dupe once https://github.com/servo/mozangle/pull/19 land.

**Comment by paulrouget** [comment]
All look green.

Need an approval for duplicating gl_generator, a r+, and we good to go.

**Pull request merged**